### ISSUE TITLE
add support for semantic naming of new tests

### ISSUE THREAD

**Issue by monperrus** [STAMP-project/dspot#647]
We could use machine learning to predict good names for new test names if the amplification is better for a new test

For this the state-the-art is "[code2vec: Learning Distributed Representations of Code](https://arxiv.org/abs/1803.09473)" and they have a open-source implementation at https://github.com/tech-srl/code2vec to start with.

**Comment by monperrus** [comment]
RW: Context2Name: A Deep Learning-Based Approach to Infer Natural Variable Names from Usage Contexts
https://arxiv.org/abs/1809.05193

**Comment by danglotb** [comment]
I looked at code2vec.

It seems quite usable and nice. First, I tried the [online demo](https://code2vec.org/), which is well documented and helps to understand code2vec.

Then, I tried to follow the example of the README. But I was not able to train the model (using the large dataset, _i.e._ 14 millions of java examples. The major issue is about the performance. I think they used GPUs to do this.

So, I wanted to try on a smaller example. I took the [test-projects](https://github.com/STAMP-project/dspot/tree/master/dspot/src/test/resources/test-projects) from the test resources of DSpot.

I managed to train and test the model. Then, I used it to predict (something, idk what yet probably a method name) of a simple amplification.

I obtained the following output:

```
Original name:	test|literal|mutation|number
	(1.000000) predicted: ['test']
Attention:
0.095443	context: METHOD_NAME,(NameExpr2)^(MethodDeclaration)_(BlockStmt)_(ExpressionStmt)_(MethodCallExpr0)_(NameExpr4),assertequals
0.086062	context: METHOD_NAME,(NameExpr2)^(MethodDeclaration)_(BlockStmt)_(ExpressionStmt)_(VariableDeclarationExpr)_(VariableDeclarator)_(VariableDeclaratorId0),ex
0.057851	context: METHOD_NAME,(NameExpr2)^(MethodDeclaration)_(BlockStmt)_(ExpressionStmt)_(VariableDeclarationExpr)_(VariableDeclarator)_(MethodCallExpr1)_(NameExpr0),ex
0.031400	context: example,(ClassOrInterfaceType0)^(VariableDeclarationExpr)_(VariableDeclarator)_(VariableDeclaratorId0),ex
0.031400	context: example,(ClassOrInterfaceType0)^(VariableDeclarationExpr)^(ExpressionStmt)^(BlockStmt)_(ExpressionStmt)_(VariableDeclarationExpr)_(VariableDeclarator)_(MethodCallExpr1)_(NameExpr0),ex
0.030653	context: void,(VoidType1)^(MethodDeclaration)_(NameExpr2),METHOD_NAME
0.026579	context: ex,(VariableDeclaratorId0)^(VariableDeclarator)^(VariableDeclarationExpr)^(ExpressionStmt)^(BlockStmt)_(ExpressionStmt)_(MethodCallExpr0)_(NameExpr4),assertequals
0.023474	context: example,(ClassOrInterfaceType0)^(VariableDeclarationExpr)^(ExpressionStmt)^(BlockStmt)_(ExpressionStmt)_(MethodCallExpr0)_(NameExpr4),assertequals
0.023474	context: example,(ClassOrInterfaceType0)^(ObjectCreationExpr)^(VariableDeclarator)^(VariableDeclarationExpr)^(ExpressionStmt)^(BlockStmt)_(ExpressionStmt)_(MethodCallExpr0)_(NameExpr4),assertequals
0.022424	context: otestliteralmutationnumber,(VariableDeclaratorId0)^(VariableDeclarator)^(VariableDeclarationExpr)^(ExpressionStmt)^(BlockStmt)_(ExpressionStmt)_(MethodCallExpr0)_(NameExpr4),assertequals
```

What I understand is, that it proposes the name `'test'` for the given input:

```java
@Test(timeout = 10000)
public void test2_literalMutationNumber9() throws Exception {
    Example ex = new Example();
    char o_test2_literalMutationNumber9__3 = ex.charAt("abcd", 0);
    Assert.assertEquals('a', ((char) (o_test2_literalMutationNumber9__3)));
}
```

I guess I did not train enough the model. In fact, the inputs is a small sets of test classes, which are artificial and test methods' names are quite similar: `test1`, `test2`, `test3` so on...

I plan to train a model using XWiki test suite (starting with one module) and retry with a better trained
model.

This first experience rises some issues/questions:
1. This process can be integrated in DSpot? It seems that to train on a large (and so to be useful)  dataset, it can take a while... I am asking myself if it is reasonable.
2. Should we use a pre-trained model? I mean, they provide models, built on top of a large number of java example but if we plan to use code2vec on test suite amplification's result, should we have a dedicated model to test or not? Should we have a model for each project or we can use one for all of them?
3. Do we have the computational power to use their approach?

Keep you updated! :-)

**Comment by danglotb** [comment]
So, I train a new model using the test suite of `xwiki-platform`, `xwiki-rendering` and `xwiki-commons.` (45 minutes of training).

Then, I tried to predict a label for the following amplified test methods:
```java
@Test 
public void explicitFilterList_literalMutationString12851_literalMutationString12935() throws Exception {
    HTMLCleanerConfiguration configuration = this.cleaner.getDefaultConfiguration();
    configuration.setFilters(Collections.emptyList());
    String result = HTMLUtils.toString(this.cleaner.clean(new StringReader("someth5ng"), configuration));
    Assertions.assertEquals("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">\n<html><head></head><body>someth5ng</body></html>\n", result);
    String String_77 = ((AmplDefaultHTMLCleanerTest.HEADER_FULL) + "<a href=\"http://xwiki.org\" target=\"_blank\">label</a>") + (AmplDefaultHTMLCleanerTest.FOOTER);
    Assertions.assertEquals("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">\n<html><head></head><body><a href=\"http://xwiki.org\" target=\"_blank\">label</a></body></html>\n", String_77);
    Assertions.assertEquals("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">\n<html><head></head><body>someth5ng</body></html>\n", result);
}
```

And obtain the following output:

```
Original name:	explicit|filter|list|literal|mutation|string|literal|mutation|string
	(0.277656) predicted: ['test', 'equalslit', 'string']
	(0.245518) predicted: ['test', 'convert', 'to', 'string']
	(0.115852) predicted: ['test', 'convert']
	(0.064386) predicted: ['restricted', 'html', 'literal', 'mutation', 'string', 'literal', 'mutation', 'string']
	(0.063261) predicted: ['explicit', 'filter', 'list', 'literal', 'mutation', 'string']
	(0.050013) predicted: ['assert', 'html']
	(0.047732) predicted: ['explicit', 'filter', 'list', 'literal', 'mutation', 'string', 'literal', 'mutation', 'string', 'literal', 'mutation', 'string']
	(0.047235) predicted: ['assert', 'html', 'with', 'head', 'content']
	(0.047199) predicted: ['clean', 'title', 'with', 'namespace']
	(0.041149) predicted: ['explicit', 'filter', 'list', 'literal', 'mutation', 'string', 'literal', 'mutation', 'string']
```

So the proposed test method name would be `'test', 'equalslit', 'string'`: `'testEqualslitString'` which to be not bad but I am not convinced.

I'm kind of puzzled here... I keep going my investigation.

**Comment by danglotb** [comment]
> Have you read anything on how to merge the name fragments together? (in which order?)

Code2vec seems to have a combinations feature. I planned to explore this ASAP.

> - In XWiki-platform, XWiki-rendering and XWiki-commons, how many test methods are there for training?
1096 for XWiki-commons, 1900 for XWiki-rendering and for XWiki-platform. In total: test methods. 

> - which tool do you use in this experiment?
What do you mean? I use code2vec, that uses tensorflow. 

But, I made a terrible mistake. I use also the source code for the training, I should use only test sources. I'm gonna work on this, and keep to explore this. Also, I do not know if I have to extract test methods or if I can five the whole test class.

I opened an issue on their GitHub's repository tech-srl/code2vec#5.

However, I'm still a bit afraid about the performance. I'm not (yet) convinced that this reasonable usable.

**Comment by urialon** [comment]
Hi guys,
Let me know if you have any more questions about code2vec.

Regarding training - we trained code2vec with a GPU, but code2vec was designed for simplicity and scale, so it does not require huge compute resources. Specifically, we trained it on a Tesla K80 (which is quite old by now) for only 30 hours. 
I think you can train a similar model using the free-tier in Azure (using an NC6 machine) or AWS (p2 machine). These VMs cost about 1$/hour and you need about 10-30 hours.
I think that both Azure&aws give 100-200$ free for students (or for anyone?).

**Comment by danglotb** [comment]
I obtained access to g5k, where there are some GPUs.

Currently, I'm training the model using the provided example of code2vec.

What I plan to do:

1. Predict a name of amplified test methods: train a "super" model using a lot (lot) of examples from GitHub. Then, specialize in this model using specific test name convention of projects under amplification.
2. Work on the minimization in order to obtain clear and readable amplified tests, that can be directly usable.
3. Work on finding a good place in the existing test suite for the amplified test: patch of an existing one or a new test method? In which test class? etc... 

The final contribution would be a post-process of DSpot to make its outputs usable, like automatize the manual minimization we did in the DSpot's paper.

**Comment by danglotb** [comment]
The training on the small dataset took: `Training time: 0H:38M:39S`.

I think this is reasonable. I'm gonna build the dataset to train on a test suite.

**Comment by monperrus** [comment]
That's a great plan!

**Comment by monperrus** [comment]
Closed per #708

**Comment by monperrus** [comment]
Automatic naming of new variables is in #830
